### ISSUE TITLE
Support/Bug: Password protected page functionality

### ISSUE THREAD

**Issue by erbanach** [timber/timber#1800]
I am using WordPress along with Timber to build a custom theme. I need a page of my website to be password protected. I have changed the settings in the WordPress admin area for this page but the out of the box WordPress functionality does not seem to work with Timber. I searched around a bit and found the following code snippets, but they do not work. What am I doing wrong here?

In my `page.php` logic:

```
    $context = Timber::get_context();
    $post = new TimberPost();
    $context['page'] = $post;
    $context['global'] = get_fields('options');
    $context['protected'] = post_password_required($post->ID); // here i am trying to check if the current page is password protected and adding the check to the context
    $context['password_form'] = get_the_password_form(); // grabbing the function to display the password form and adding it to the context
    Timber::render( array(
    'pages/page-' . $post->post_name . '.twig',
    'pages/page.twig'
    ), $context );
```

In my Twig template, I tried:

```
    {% if protected %}
    {{ password_form }}
    {% else %}
    <!-- rest of page template here -->
    {% endif %}
```

This `if` check does not seem to work, it always says that the page is not password protected even if it has been specified otherwise in the WP admin. The `{{ password_form }}` does display the password form correctly. How can I get this working?

I also tried this in `page.php`:

```
    $context = Timber::get_context();
    $post = new TimberPost();
    $context['page'] = $post;
    $context['global'] = get_fields('options');

    if ( post_password_required($post->ID) ) {
      Timber::render('page-password.twig', $context);
      } else {
      Timber::render(array('pages/page-' . $post->post_name . '.twig', 'pages/page.twig' ), $context);
      }
```

With `page-password.twig` displaying the password form. This if check also did not work. This might be a Timber bug because I believe those functions should be working.

[Stack overflow question here.](https://stackoverflow.com/questions/52505820/password-protected-page-functionality-not-working-out-of-the-box-with-timber-twi)

**Comment by marciojc** [comment]
@erbanach you are seeing the page as admin? you must be logged out to see the form.

**Comment by jarednova** [comment]
saw that there's a good amount of activity over on StackOverflow, moving there ...

https://stackoverflow.com/questions/52505820/password-protected-page-functionality-not-working-out-of-the-box-with-timber-twi

**Comment by gchtr** [comment]
No matter what the final solution for this will be, I think we could add a section in the documentation for how to do password protected pages in Timber, maybe in the [WordPress Integration Guide](https://timber.github.io/docs/guides/wp-integration/)?

**Comment by tnottu** [comment]
I've thought that it is quite unsafe and unpractical to do this for every template/view individually. If you forget to make the check in some template (and then try to password-protect a page using it), you might be in trouble. Also, you have to repeat code, perhaps unnecessarily.

My suggestion is to use a specific php-template for all password protected pages. Here is the filter I have used:

```php
/**
 * Use specific template for password protected posts
 */

add_filter( 'template_include', 'get_password_protected_template', 99 );

function get_password_protected_template( $template ) {

  global $post;

  if ( !empty( $post ) && post_password_required( $post->ID ) ) {
    $template = locate_template( ['password-protected.php'] ) ?: $template;
  }

  return $template;
}
```

And then I have a `password-protected.php` and `password-protected.twig` template and view to show a login page with the form.

**Comment by gchtr** [comment]
@tnottu I like you solution very much 😎! This is indeed much cleaner. And even if you wanted to use different templates for different post types, you could do:

```php
$template = locate_template( [
    'password-protected.php',
    "password-protected-{$post->post_type}.php",
] ) ?: $template;
```

I added a [pull request](https://github.com/timber/timber/pull/1804) that adds your solution to the documentation.

**Comment by jarednova** [comment]
Closed by #1804

**Comment by terrance-orletsky-d7** [comment]
Just as a warning to some others - while this solution may work for some, in my Timber/Twig theme it caused a lot of havok by hijacking the query too soon, and causing archive pages to be password protected if they contained a password post.  In the end, I just reverted to the theme based approach.  Hope this helps someone, and thanks for sharing your approach!
